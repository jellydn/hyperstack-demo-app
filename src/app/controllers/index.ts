// naming and imports: name and import as you see fit
import Auth from './auth'
import { NotesController } from './notes'

const controllers = []
controllers.push(Auth)
controllers.push(NotesController)

export const AppController = {
  controllers,
}
