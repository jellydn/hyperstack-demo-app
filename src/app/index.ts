import type { Context } from 'hyperstack'
import { context } from 'hyperstack'
import type * as Models from './models'
import type * as Mailers from './mailers'

export const appContext = context as Context<typeof Models, typeof Mailers>
