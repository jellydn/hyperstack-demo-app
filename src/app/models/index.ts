export { Note } from './note'
export { User } from './user'
