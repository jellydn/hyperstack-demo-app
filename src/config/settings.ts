// general purpose settings, with no infrastructure loading, no assumptions. keep it simple.
import path from 'path'
export const root = path.join(__dirname, '..')
