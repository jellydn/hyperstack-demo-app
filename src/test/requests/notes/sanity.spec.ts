import { test } from '@hyperstackjs/testing'
import L from 'lodash'
import { root } from '../../../config/settings'
import { appContext } from '../../../app'

const {
  requests,
  matchers: { matchRequestWithSnapshot },
} = test(root)

const pid = /.+-.+-.+-.+-.+/
const serializeNote = (note: any) => {
  expect(note.pid).toMatch(pid)
  note.pid = 'test-redacted'
  return note
}
const serializeOneNote = (res: any) => {
  res.body.note = serializeNote(res.body.note)
  return res
}
const serializeNotes = (res: any) => {
  res.body.notes = L.map(res.body.notes, serializeNote)
  return res
}

const auth = (token: string) => ({ Authorization: `Bearer ${token}` })
describe('requests', () => {
  describe('/notes', () => {
    const getToken = async () => {
      const { User } = appContext.models()
      const user = await User.createWithPassword({
        username: 'evh@example.com',
        password: 'mypass-should-login',
        name: 'Eddie Van Halen',
      })
      return { token: user.createAuthenticationToken(), user }
    }
    requests('should list', async (request, _app, _server) => {
      const { token, user } = await getToken()
      const { Note } = appContext.models()
      await Note.createWithOwner(user, {
        content: 'interesting note',
        title: 'this is the title',
      })

      await matchRequestWithSnapshot(
        200,
        request().get('/notes').set(auth(token)),
        { serializer: serializeNotes }
      )
    })
    requests('should get one', async (request, _app, _server) => {
      const { token, user } = await getToken()
      const { Note } = appContext.models()
      const note = await Note.createWithOwner(user, {
        content: 'interesting note',
        title: 'this is the title',
      })

      await matchRequestWithSnapshot(
        200,
        request().get(`/notes/${note.pid}`).set(auth(token)),
        { serializer: serializeOneNote }
      )
    })
    requests('should create', async (request) => {
      const { token } = await getToken()
      await matchRequestWithSnapshot(
        200,
        request()
          .post(`/notes`)
          .set(auth(token))
          .send({ title: 'new note', content: 'dont ask dont know' }),
        { serializer: serializeOneNote }
      )
    })
    requests('should update', async (request) => {
      const { token, user } = await getToken()
      const { Note } = appContext.models()
      const note = await Note.createWithOwner(user, {
        content: 'interesting note',
        title: 'this is the title',
      })

      await matchRequestWithSnapshot(
        200,
        request()
          .post(`/notes/${note.pid}`)
          .set(auth(token))
          .send({ title: 'boring note', content: 'boring content' }),
        { serializer: serializeOneNote }
      )
    })
    requests('should be secure', async (request, _app, _server) => {
      const { user } = await getToken()
      const { Note, User } = appContext.models()
      const mallory = await User.createWithPassword({
        username: 'mallory@example.com',
        password: 'give-me-all-your-stuff',
        name: 'Mallory Vicious',
      })
      const note = await Note.createWithOwner(user, {
        content: 'interesting note',
        title: 'this is the title',
      })

      // no auth
      await matchRequestWithSnapshot(401, request().get(`/notes`))

      // mallory snoops another user's note
      await matchRequestWithSnapshot(
        404,
        request()
          .get(`/notes/${note.pid}`)
          .set(auth(mallory.createAuthenticationToken()))
      )

      // mallory updates another user's note
      await matchRequestWithSnapshot(
        404,
        request()
          .post(`/notes/${note.pid}`)
          .set(auth(mallory.createAuthenticationToken()))
      )
    })
  })
})
