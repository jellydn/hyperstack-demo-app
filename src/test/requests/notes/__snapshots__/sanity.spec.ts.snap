// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requests /notes should be secure 1`] = `
{
  "body": {},
  "headers": {
    "connection": "close",
    "content-length": "8",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "text/html; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "www-authenticate": "",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 401,
}
`;

exports[`requests /notes should be secure 2`] = `
{
  "body": {
    "error": "note not found",
  },
  "headers": {
    "connection": "close",
    "content-length": "26",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 404,
}
`;

exports[`requests /notes should be secure 3`] = `
{
  "body": {
    "error": "note not found",
  },
  "headers": {
    "connection": "close",
    "content-length": "26",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 404,
}
`;

exports[`requests /notes should create 1`] = `
{
  "body": {
    "note": {
      "content": "dont ask dont know",
      "pid": "test-redacted",
      "title": "new note",
    },
  },
  "headers": {
    "connection": "close",
    "content-length": "105",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 200,
}
`;

exports[`requests /notes should get one 1`] = `
{
  "body": {
    "note": {
      "content": "interesting note",
      "pid": "test-redacted",
      "title": "this is the title",
    },
  },
  "headers": {
    "connection": "close",
    "content-length": "112",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 200,
}
`;

exports[`requests /notes should list 1`] = `
{
  "body": {
    "notes": [
      {
        "content": "interesting note",
        "pid": "test-redacted",
        "title": "this is the title",
      },
    ],
  },
  "headers": {
    "connection": "close",
    "content-length": "115",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 200,
}
`;

exports[`requests /notes should update 1`] = `
{
  "body": {
    "note": {
      "content": "boring content",
      "pid": "test-redacted",
      "title": "boring note",
    },
  },
  "headers": {
    "connection": "close",
    "content-length": "104",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 200,
}
`;
